File: lintcode/543.kth-largest-in-n-arrays.cpp
class Solution {
public:
    /**
     * @param arrays: a list of array
     * @param k: An integer
     * @return: an integer, K-th largest element in N arrays
     */
    int KthInArrays(vector<vector<int>> &arrays, int k) {
        // write your code here
        // default is max queue, use greater comparator to make a min queue
        priority_queue<int, vector<int>, greater<int>> pq; 

        for (auto list: arrays)
        {
            for (auto num: list)
            {
                if (pq.size() < k)
                {
                    pq.push(num);
                }
                else if (pq.top() < num)
                {
                    pq.pop();
                    pq.push(num);
                }
            }
        }

        return pq.top();
    }
};

File: lintcode/543.kth-largest-in-n-arrays.py
import heapq
class Solution:
    """
    @param arrays: a list of array
    @param k: An integer
    @return: an integer, K-th largest element in N arrays
    """
    def KthInArrays(self, arrays, k):
        # write your code here
        hq = []

        for sub in arrays:
            for num in sub:
                if len(hq) < k:
                    heapq.heappush(hq, num)
                elif hq[0] < num:
                    heapq.heappop(hq)
                    heapq.heappush(hq, num)

        return hq[0]